lib/rspec/core/reporter.rb

```diff
@@ -79,6 +88,22 @@ def message(message)
       notify :message, Notifications::MessageNotification.new(message)
     end
 
+    # @param event [Symbol] Name of the custom event to trigger on formatters
+    # @param options [Hash] Hash of arguments to provide via `CustomNotification`
+    #
+    # Publish a custom event to supporting registered formatters.
+    # @see RSpec::Core::Notifications::CustomNotification
+    def publish(event, options={})
+      raise UnsupportedPublish if RSPEC_NOTIFICATIONS.include? event
+      notify event, Notifications::CustomNotification.for(options)
+    end
+
+    # @private
+    UnsupportedPublish = StandardError.new(
+      "RSpec::Core::Reporter#publish is intended for sending custom " \
+      "events not internal RSpec ones, please rename your custom event."
+    )
```

Comment:
This is kinda non-standard for how we've normally done errors.  `UnsupportedPublish` looks like it's a class constant but it's not.  Also, using a singleton instance is potentially problematic if the message gets mutated...

What advantage did you see to doing it this way?

BTW, I think it should use `ArgumentError`, not `StandardError` as it's for a bad argument, right?